Various optimizations to AMP inabox friendly iframe position observer.

zombifier opened the pull request:
1. If `window.frameElement` is not available (e.g. because the ad IS the top window) use `getScrollingElement()`. This is mostly relevant for mobile apps.

2. Both the host script and the amp4ads runtime within friendly iframes will now share one position observer (sitting at the host window) instead of using a separate one for each. This prevents multiple scroll listeners, which can reduce performance.

3. Fix single pass build, which was actually broken before because of minification - the position observer would send back the minified `data` object while the inabox runtime would still try to read the unminified name (`data['viewportRect']`). The object will now be unminified in all cases.

lannka commented on ads/inabox/position-observer.js:
```diff
@@ -179,3 +179,15 @@ function getScrollingElement(win) {
+ * @param {!Window} win
+ * @return {!PositionObserver}
+ */
+export function getPublicPositionObserver(win) {
```
nit: `getPositionObserver`

lannka commented on ads/inabox/position-observer.js:
```diff
@@ -179,3 +179,15 @@ function getScrollingElement(win) {
+ * @return {!PositionObserver}
+ */
+export function getPublicPositionObserver(win) {
+  const AMP_INABOX_POSITION_OBSERVER = 'ampInaboxPositionObserver';
```
nit: move constant out to the top

zombifier commented:
Addressed review, but don't merge yet. The `PositionObserver` class and `getPositionObserver()` is compiled into two different binaries, and I'm trying to confirm that the minified class when stored in the window variable will be interpreted the same by both `amp4ads-v0.js` and `amp4ads-host-v0.js`. Testing manually has not shown anything wrong so far, but I'll be doing some code inspection to be sure.

zombifier commented:
OK, the names of private variables are different as expected but the object prototype that contains the function is the same, so there's no problem on that front.

lannka commented:
SG. need to keep an eye on it for single pass compilation .

zombifier commented:
So in single pass mode the public methods of `PositionObserver` are not exported to the prototype if they are not immediately used (e.g. `amp4ads-v0.js` will not export `getViewportRect()` when it creates the position observer, even if the host script `amp4ads-host-v0.js` uses it). What's the standard way of preventing this? `@export` doesn't seem to work.

Alternatively I could have `amp4ads-host-v0.js` use its own position observer again so we don't have to worry about multiple binaries.

zombifier commented:
@lannka I made the required changes, single pass should work now. Can you take another look and restart Travis?

The single pass problem actually persists before this PR in a different way (see my edit in the PR description); the tests didn't catch it since none of the tested browsers lack IntersectionObserver, which is sufficient for detecting visibility.

lannka commented on ads/inabox/position-observer.js:
```diff
@@ -92,9 +95,9 @@ export class PositionObserver {
   getPositionEntry_(element) {
     return {
-      viewportRect: /** @type {!LayoutRectDef} */ (this.viewportRect_),
+      'viewportRect': /** @type {!LayoutRectDef} */ (this.viewportRect_),
```
thanks for the fix!

lannka commented on ads/inabox/position-observer.js:
```diff
@@ -179,3 +182,14 @@ function getScrollingElement(win) {
+ * @param {!Window} win
+ * @return {!PositionObserver}
+ */
+export function getPositionObserver(win) {
```
this seems only needed in host.js? we can move it over there.

zombifier replied:
It's used in `src/inabox/inabox-viewport.js` as well. It could be kept here.

lannka commented on src/inabox/inabox-viewport.js:
```diff
@@ -198,10 +198,10 @@ export class ViewportBindingInabox {
       this.unobserveFunction_ = this.topWindowPositionObserver_.observe(
-        /** @type {!HTMLIFrameElement} */ (this.win.frameElement),
+        /** @type {!HTMLIFrameElement|!HTMLElement} */
+        (this.win.frameElement || this.getScrollingElement()),
```
pls document when this can be `null`

zombifier replied:
Done.
